TEE.md

```diff
@@ -0,0 +1,79 @@
+2. The node encrypts the sensitive secret shares with the derived key.
+3. The node exports the encrypted secret shares.
+
+#### Agreeing on an encryption key
+A pair-wise key establishement scheme can be leveraged to establish a symmetric key. A suitable implementation can be found in [libsodium](https://doc.libsodium.org/) and its rust wrapper [libsodium-rs](https://docs.rs/crate/libsodium-rs/latest). Specifically, the **key exchange protocol based on X25519** would be a good fit for the given set-up (c.f. [libsodium](https://doc.libsodium.org/key_exchange) and [libsodium-rs](https://docs.rs/libsodium-rs/latest/libsodium_rs/crypto_kx/index.html)).
+
+_Note: This key exchange protocol is on Curve25519, which is different from the Edwards curve used by NEAR._
+
+The key establishment scheme requires that the node as well as the back-up recipient have *mutually authenticated Curve25519 public keys*. The NEAR blockchain can be leveraged for this, more specifically:
+- the node operator whitelists a NEAR-account associated to the back-up recipient in the MPC smart contract.
+- the back-up recipient submits a public key (Curve25519) to the MPC smart contract. The smart contract only accepts it if the account has been whitelisted by the node operator.
+- the node submits a public key (Curve25519) to the MPC smart contract. The smart contract only accepts it if the public key submitting this transaction is an access key of the node operators account.
+
+_Note: Curve25519 and Edwards curve are [birationally equivalent](https://crypto.stackexchange.com/questions/43013/what-does-birational-equivalence-mean-in-a-cryptographic-context), so one could theoretically use NEAR account keys. But it seems cleaner if we just have the node and the back-up recipient generate dedicated keys for this._
+
+#### Encrypting the secret
+The node running inside the TEE monitors the contract and can derive an encryption key using its secret key and the public key submitted by the back-up recipient.
+
+#### Exporting the encrypted secret
+The back-up service triggers the export of the back-up through a call to the smart contract or a dedicated web-endpoint.
+The node returns the encrypted key either:
+- as a response to a request to its web-endpoint;
+- by printing its hex-value to the logs;
+- or through some other mechanism (@Barak, are there alternatives?)
+
+The back-up mechanism can then decrypt the received ciphertext and store the secret somewhere safe.
+
+_Todo: improve this section with more details. It is vague._
+
+### Restoration Mechanism:
+#### Encryption Key Agreement & Encrypting the secret 
+Same as above, just reversed roles
+
+#### Importing the encrypted secret into the node
+Similar to above, but reversed roles. The node makes a web request to the back-up service. (@Barak is this possible?)
+_Todo: improve this section with more details. It is vague._
```

Comment:
I don't see why this shouldn't be possible.

However, it would be nice if the backup service didn't expose any HTTP endpoint but instead the backup service posts the data to an endpoint on the node. This way, the node doesn't have to be aware of the backup service, making for a more loosely coupled design.

This also allows for an initial design where the backup service is just a simple CLI saving the shares to disk, which perhaps could be a good-enough MVP for the initial soft-launch.

Reply:
>  it would be nice if the backup service didn't expose any HTTP endpoint but instead the backup service posts the data to an endpoint on the node. 

Oh yeah, that would be way nicer!